crates/bevy_asset/src/asset_changed.rs

```diff
@@ -0,0 +1,451 @@
+//! Define the [`AssetChanged`] query filter.
+//!
+//! Like [`Changed`], but for [`Asset`]s.
+
+use bevy_ecs::{
+    archetype::Archetype,
+    component::{ComponentId, Tick},
+    prelude::{Entity, Resource, World},
+    query::{FilteredAccess, QueryFilter, QueryItem, ReadFetch, WorldQuery},
+    storage::{Table, TableRow},
+    world::unsafe_world_cell::UnsafeWorldCell,
+};
+use bevy_utils::HashMap;
+use disqualified::ShortName;
+use core::marker::PhantomData;
+
+use crate::{AsAssetId, Asset, AssetId};
+
+#[derive(Resource)]
+pub(crate) struct AssetChanges<A: Asset> {
+    change_ticks: HashMap<AssetId<A>, Tick>,
+    last_change_tick: Tick,
+}
+
+impl<A: Asset> AssetChanges<A> {
+    pub(crate) fn insert(&mut self, asset_id: AssetId<A>, tick: Tick) {
+        self.last_change_tick = tick;
+        self.change_ticks.insert(asset_id, tick);
+    }
+    pub(crate) fn remove(&mut self, asset_id: &AssetId<A>) {
+        self.change_ticks.remove(asset_id);
+    }
+}
+
+impl<A: Asset> Default for AssetChanges<A> {
+    fn default() -> Self {
+        Self {
+            change_ticks: Default::default(),
+            last_change_tick: Tick::new(0),
+        }
+    }
+}
+
+struct AssetChangeCheck<'w, A: AsAssetId> {
+    change_ticks: &'w HashMap<AssetId<A::Asset>, Tick>,
+    last_run: Tick,
+    this_run: Tick,
+}
+
+impl<A: AsAssetId> Clone for AssetChangeCheck<'_, A> {
+    fn clone(&self) -> Self {
+        *self
+    }
+}
+
+impl<A: AsAssetId> Copy for AssetChangeCheck<'_, A> {}
+
+impl<'w, A: AsAssetId> AssetChangeCheck<'w, A> {
+    fn new(changes: &'w AssetChanges<A::Asset>, last_run: Tick, this_run: Tick) -> Self {
+        Self {
+            change_ticks: &changes.change_ticks,
+            last_run,
+            this_run,
+        }
+    }
+    // TODO(perf): some sort of caching? Each check has two levels of indirection,
+    // which is not optimal.
+    fn has_changed(&self, handle: &A) -> bool {
+        let is_newer = |tick: &Tick| tick.is_newer_than(self.last_run, self.this_run);
+        let id = handle.as_asset_id();
+
+        self.change_ticks.get(&id).is_some_and(is_newer)
+    }
+}
+
+/// Filter that selects entities with a `A` for an asset that changed
+/// after the system last ran, where `A` is a component that implements
+/// [`AsAssetId`].
```

Comment:
What about components that store handles to multiple assets of the same type? We don't need to do that right now, but we might have to revisit it later. In my app I have components with handles to dozens of `AnimationClip`s for example. I think you could maybe just have `AsAssetId`'s method return `-> impl Iterator<Item = AssetId>`.

This isn't blocking, just a suggestion.

Reply:
The bigger issue is how to handle the associated type. We might have to use some kind of indirection via UntypedAssetId instead to be able to handle multiple kinds of assets.

Reply:
Sorry, I also misunderstood you. Multiple assets of the same type should be much easier to handle. I'd been hung up on if, for example we decided to merge mesh and material as has been discussed before.